### PR TITLE
Issue/deadlock on boot

### PR DESCRIPTION
**Pull request by wouterdb**
# Description

Fix deadlock on server boot

This is a bit subtle because:
1. lsm validation listener is blocking 
2. lsm validation listener calls request compile

So the process has to be non-blocking both ways (boot can not block new requests, listeners can not block boot)  and order preserving (listener goes before compile, compiles go in order)

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [ ] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

File: src/inmanta/server/services/compilerservice.py [hunk 1]
```diff
@@ -515,9 +515,14 @@ class CompilerService(ServerSlice, environmentservice.EnvironmentListener):
     def __init__(self) -> None:
         super().__init__(SLICE_COMPILER)
         self._recompiles: dict[uuid.UUID, Task[None | Result]] = {}
+
         self._global_lock = asyncio.locks.Lock()
+        # boolean indicating when everything is up and processing can start
```
Comment (sanderr) [inline src/inmanta/server/services/compilerservice.py:520]:
```suggestion
        # boolean indicating when everything is up and compile execution can start
```
If I understand correctly this determines only if we are ready to start executing compiles, not whether the service is ready to receive requests or anything like that. Right?

**Review by sanderr — APPROVED** [review]

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/server/services/compilerservice.py (hunk 1 above)
Comment (wouterdb) [reply src/inmanta/server/services/compilerservice.py]: correct

File: src/inmanta/server/services/compilerservice.py [hunk 2]
```diff
@@ -774,15 +781,23 @@ async def _sub_notify() -> None:
...
+        async def sub_recovery() -> None:
+            unhandled = await data.Compile.get_unhandled_compiles()
+            for u in unhandled:
+                await self._notify_listeners(u)
+            # We are ready to start!
+            self.fully_ready = True
+            # one run per env max to get started
+            runs = await data.Compile.get_next_run_all()
+            for run in runs:
+                await self._queue(run)
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/server/services/compilerservice.py:798]: thx, missed that one

[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/server/services/compilerservice.py:798]: Could you elaborate? `_queue` does not actually schedule the compile if another is already in progress.

[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/server/services/compilerservice.py]: replaced

File: src/inmanta/server/services/compilerservice.py [hunk 3]
```diff
@@ -774,15 +781,44 @@ async def _sub_notify() -> None:
...
+
+
+
+            Must be called under the _global_lock
+        """
+        if self.is_stopping():
+            return
+        env: Optional[data.Environment] = await data.Environment.get_by_id(environment)
+        if env is None:
+            LOGGER.warning("Can't dequeue compile: environment %s does not exist" % environment)
```
Comment (arnaudsjs) [inline src/inmanta/server/services/compilerservice.py:817]:
```suggestion
            LOGGER.warning("Can't dequeue compile: environment %s does not exist", environment)
```

File: src/inmanta/server/services/compilerservice.py [hunk 4]
```diff
@@ -774,15 +781,44 @@ async def _sub_notify() -> None:
...
+            # prevent races by holding global lock
+            async with self._global_lock:
+                # We are ready to start!
+                self.fully_ready = True
+                for run in runs:
+                    await self.__queue_no_lock(run.environment, run)
+
+        self.add_background_task(sub_recovery())
+
+    async def __queue_no_lock(self, environment: uuid.UUID, compile: data.Compile) -> None:
```
Comment (arnaudsjs) [inline src/inmanta/server/services/compilerservice.py:805]: Is there a reason why `_queue()` doesn't call into this method? Now this code is duplicates across two methods.

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

**Review by arnaudsjs — APPROVED** [review]

**Pull request closed without merging**